src/util-macset.c

```diff
@@ -0,0 +1,120 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+/**
+ * \file
+ *
+ * \author Sascha Steinbiss <sascha.steinbiss@dcso.de>
+ *
+ * Set-like data store for MAC addresses. Implemented as array for memory
+ * locality reasons as the expected number of items is typically low.
+ *
+ */
+
+#include "suricata-common.h"
+#include "suricata.h"
+#include "util-macset.h"
+
+typedef uint8_t MacAddr[6];
+
+struct MacSet_ {
+    MacAddr *buf[2];
+    unsigned long size,
+                  last[2];
+};
+
+MacSet* MacSetInit(int size)
```

Comment:
I'm not too excited for having 3 memory allocations here for the common case. Could we have just 1 that stores a single src and dest mac, and only do additional allocs in uncommon case of having multiple macs per flow side?

Reply:
I was aware of that, but noticing that flow data structures are apparently recycled (reducing the amount of new allocations) went for simplicity rather than allocation efficiency. Sure, I'll look into that, at the cost of some additional complexity.